### ISSUE TITLE
Support for temporary reply-to queues

### ISSUE THREAD

**Issue by weibeld** [stomp-js/ng2-stompjs#85]
I'm using RabbitMQ 3.7.7 as a broker, and I'm trying to use a temporary `reply-to` queue by specifying `/temp-queue/foo` as the value of the `reply-to` header, as described [here](https://medium.com/@gottfrois/temporary-queues-with-rabbitmq-using-web-stomp-e98e266d61e1) or [here](https://activemq.apache.org/apollo/documentation/stomp-manual.html#Temporary_Destinations):

~~~
this._stompService.publish('/queue/request-queue', 'Test message', {'reply-to': '/temp-queue/foo'});
~~~

This seems actually to work fine, as an auto-named temporary queue is indeed created on RabbitMQ and the `StompService` receives the response from the server on this queue, without even the need to manually subscribe to this queue:

![screen shot 2018-09-20 at 11 42 46](https://user-images.githubusercontent.com/2396208/45811299-00a60d00-bccd-11e8-8345-5871eb35713c.png)

This would all be great, but how can I get the reponse message in my client application from the `StompService`? I didn't find a way to provide a callback for messages that are received in this way.

For example, it's not possible to subscribe to this queue like this:

~~~
this._stompService.subscribe('/temp-queue/foo');
~~~

This results in an error:

![screen shot 2018-09-20 at 11 52 52](https://user-images.githubusercontent.com/2396208/45811554-abb6c680-bccd-11e8-9da8-cd6926ab9048.png)

**Comment by kum-deepak** [comment]
Great! I was waiting for someone to raise this question. 

I have been experimenting with temp queues for purpose of async RPC (in the rpc branch).

Which version of this library are you using (4.x.x or 6.x.x.)?

**Comment by kum-deepak** [comment]
Great, give me few days. I will upgrade the code (that part of the code is still for Angular 4/5). it does work with Angular4/5. It has reasonable test coverage as well. I wanted someone to run the API and the model with someone before I merged it to released version.

The way I have been planning is to automatically populate a co-relation-id and route the response to the caller.

**Comment by weibeld** [comment]
Great, looking forward to the change. Let me know if you need any more information!

**Comment by kum-deepak** [comment]
Just published `6.1.0-beta.1` at npm. This is from `rpc-v6` branch.

Following is how it works:

- There is [defaultMessagesObservable](https://stomp-js.github.io/ng2-stompjs/injectables/StompRService.html#defaultMessagesObservable) - this will trigger on any message that is sent to any of the temporary queue.
- The `StompRPCService` has a call `rpc` with following signature:

```typescript
rpc(serviceEndPoint: string, payload: string): Observable<Message>
```
- The returned Observable will trigger once with the response.
- When making an rpc call, I am creating a new UUID and pass it as `correlation-id` as message header. Temporary queue name is set as `reply-to` header. Currently I am using same temp queue name for all calls.
- The message is sent to the required end point. The endpoint - after processing - is supposed to publish the response to the `reply-to` queue and add same `correlation-id` as header.

Please see the code at https://github.com/stomp-js/ng2-stompjs/blob/rpc-v6/src/stomp-rpc.service.ts

Test cases at https://github.com/stomp-js/ng2-stompjs/blob/rpc-v6/src/specs/app/services/stomp-rpc.service.spec.ts

See if the above makes sense. If you need please copy the code into your project and test out what works for you.

I have carried out all of my testing using RabbitMQ.

**Comment by weibeld** [comment]
I tested it, works perfectly!

I didn't run the test cases, but I tested it in a test application with RabbitMQ and an RPC server written in Python with [pika](https://pika.readthedocs.io/en/latest/) (using AMQP) on the other end. Extraction of the correlation ID and reply-to queue on the server side works perfectly, and the `StompRPCService` receives the response on the temporary queue and resolves the `Observable` only when the correlation IDs match. So, everything works as it should.

I have just one proposition, would it be possible to still allow a headers argument int the `rpc` method:

~~~
rpc(serviceEndPoint: string, payload: string, headers: StompHeaders)
~~~

Because otherwise, with the default settings, the RPC queue must be durable and non-auto-delete, but it would be good to be able to use it also with different types of queues, for example:

~~~
rpc('/queue/request-queue', 'Test message', {durable: false, 'auto-delete': true})
~~~

On another note, I still had to run `npm install angular2-uuid --save` to make it compile.

**Comment by kum-deepak** [comment]
Read the activemq documentation, the code that I have written will not work.

RabbitMQ STOMP adapter (at the broker end) automatically subscribes to `reply-to` queue. However in case of ActiveMQ we need to explicitly subscribe to the temporary queue before making our request.

Unfortunately this part of the functionality is part of STOMP standards to these brokers have implemented it differently. 

Please let me know if I need to put sample code that should work for ActiveMQ.

**Comment by kum-deepak** [comment]
Thanks for your testing and feedback. Your suggestions are quite useful. Pushed another tag `6.1.0-beta.2`. Changes:

- Allow headers for rpc requests.
- The service now needs `StompRService` instead of `StompService` as dependency.
- Moved `angular2-uuid` from peerDependencies to dependencies.

In RabbitMQ we do not need to subscribe explicitly to the temp queue. So, we do not need headers for queue subscription.

The current code should be sufficient for RabbitMQ.

For ActiveMQ the code will need to be slightly refactored. I will be trying out few approaches after I make a dev setup of ActiveMQ.

**Comment by kum-deepak** [comment]
Many thanks, I wrote first version in June 2017. I was waiting for someone to show interest. 👍 

Pushed one more beta. It should work out of the box with RabbitMQ. There are instructions for other brokers. Added a guide: https://github.com/stomp-js/ng2-stompjs/blob/rpc-v6/docs-src/RPC.md

Based on your feedback I will merge and release it.

**Comment by weibeld** [comment]
Just tested version `6.1.0-beta.3`. I encounter the following problem with `StompRPCConfig`:

In `app.module.ts`, my IDE (WebStorm) completes the import of `StompRPCConfig` automatically to this:

~~~typescript
import {StompRPCConfig} from '@stomp/ng2-stompjs/src/stomp-rpc.config';
~~~

In contrast to the other classes which can be imported like this:

~~~typescript
import {StompRPCService} from '@stomp/ng2-stompjs';
~~~

Then I have the following in `app.module.ts`:

~~~typescript
const stompRPCConfig: StompRPCConfig = {};

@NgModule({
  // ...
  providers: [
    // ...
    {
      provide: StompRPCConfig,
      useValue: stompRPCConfig
    }
  ]
})
export class AppModule {}
~~~

And when I compile, I get the following compiler error:

~~~
ERROR in ./src/app/app.module.ts
Module not found: Error: Can't resolve '@stomp/ng2-stompjs/src/stomp-rpc.config' in '/Users/dw/Desktop/ng6-stompjs-demo/src/app'
~~~

Do you know what's the problem here?

**Comment by weibeld** [comment]
Or maybe better rephrased, when I do the following in `app.module.ts`:

~~~typescript
import { StompRPCConfig } from '@stomp/ng2-stompjs';
~~~

I get the following TSLint and compiler error:

~~~
TS2305: Module '"....../node_modules/@stomp/ng2-stompjs/stomp-ng2-stompjs"' has no exported member 'StompRPCConfig'.
~~~

**Comment by kum-deepak** [comment]
It was an packaging error. I needed to declare the newly added class to be part of the distribution package. Pushed one more beta, see if this fixes the issue 😄

**Comment by weibeld** [comment]
This issue is fixed with `6.1.0-beta.4`, thanks. But now I get another error at runtime.

When I call the `rpc` method, I get the following:

![screen shot 2018-09-24 at 12 01 04](https://user-images.githubusercontent.com/2396208/45946869-84743800-bff2-11e8-99a7-125ea2a067d2.png)

And the code section that it refers to is this one:

![screen shot 2018-09-24 at 12 03 31](https://user-images.githubusercontent.com/2396208/45946914-a8377e00-bff2-11e8-8a43-9bd8b7facbed.png)

I never had this error in the previous versions.

**Comment by weibeld** [comment]
I just tried it, it doesn't make a difference. There is the same error message occurring.

**Comment by weibeld** [comment]
I did some experiments with an own version of the `StompRPCService` class. If this class takes in the constructor a `StompRService` (as it's the case in the current `6.1.0-beta.4` version), then I get exactly the error from above. However, if I change the constructor argument to `StompService`, then everything works correctly.

I think you [changed](https://github.com/stomp-js/ng2-stompjs/issues/85#issuecomment-423589685) the constructor argument for `StompRPCService` from `StompService` to `StompRService` in `6.1.0-beta.2`, and I remember that I didn't get this error when I tested `6.1.0-beta.1`.

So, could this be the problem? If you push a new version with just this constructor argument changed, then I could test it.

**Comment by kum-deepak** [comment]
`StompService` is derived from `StompRService` so, would you try the following:

```typeacript
{ provide: StompRService, useExisting: StompService },
```

Based on your results, I will make further changes.

Today going through https://angular.io/guide/dependency-injection-in-action I also realized that in `StompRPCService`, I need to mark config as `@Optional()`.

**Comment by weibeld** [comment]
With this in change it works.

However, now when a message is received by my RPC server, implemented in Java, the server channel is closed due to this [error](https://stackoverflow.com/questions/52537258/rabbitmq-find-reason-for-channel-shutdown-due-to-consumer-exception). The same RPC server works with other RPC clients. I didn't yet find out what the problem is.

If I use it with another RPC server implemented in Python, then it works.

**Comment by weibeld** [comment]
Btw, I found another error with the `headers` argument. At the moment, the `headers` argument to `rpc` and `stream` (lines [41](https://github.com/stomp-js/ng2-stompjs/blob/75428940ae69bc28b48674c247c98edd80fce6ad/src/stomp-rpc.service.ts#L41) and  [49](https://github.com/stomp-js/ng2-stompjs/blob/75428940ae69bc28b48674c247c98edd80fce6ad/src/stomp-rpc.service.ts#L49)) is used directly as an argument to `publish` (line [70](https://github.com/stomp-js/ng2-stompjs/blob/75428940ae69bc28b48674c247c98edd80fce6ad/src/stomp-rpc.service.ts#L70)).

However, this same `headers` variable seems to be reused for subsequent calls to `rpc`. With my reimplementation, when I made two `rpc` calls programmatically right after another, then the second call always overwrote the `headers` object of the first call, before the first call completed its `publish` operation. So, both messages were eventually sent with the `headers` object of the second call, that means, they also used the same correlation ID.

Creating a copy of the `headers` object solved this problem for me:

~~~
const headersCopy = {};
for (let key in headers) {
  headersCopy[key] = headers[key];
}

headersCopy['reply-to'] = this._replyQueueName;
headersCopy['correlation-id'] = correlationId;

this.stompService.publish(serviceEndPoint, payload, headersCopy);
~~~

It also works if `headersCopy` is created outside of the observable definition, so maybe it could be the first thing to be done in the `stream` method.

**Comment by kum-deepak** [comment]
Thanks! Indeed it is a bug. I will switch to use `Object.assign` which is supported by most of the browsers and polyfill exists for others.

I will inspect the code for `stompjs` (the underlying library) for same issue - it is likely to be there as well.

**Comment by weibeld** [comment]
Hi, do you have any plans when v7 will be ready to release?

**Comment by kum-deepak** [comment]
In couple of weeks 😄

**Comment by kum-deepak** [comment]
Will be released as 7.0.0
